Plugin/Magento/Fedex/Model/CacheFedexResultPlugin.php

```diff
@@ -0,0 +1,166 @@
+    public function __construct(
+        Bugsnag $bugsnag,
+        CacheInterface $cache,
+        SessionHelper $sessionHelper,
+        Json $json = null
+    ) {
+        $this->bugsnag = $bugsnag;
+        $this->sessionHelper = $sessionHelper;
+        $this->cache = $cache;
+        $this->json = $json ?: ObjectManager::getInstance()->get(Json::class);
+        // Call protected method with a Closure proxy
+        $this->methodCaller = function ($methodName, ...$params) {
+            return $this->$methodName(...$params);
+        }; 
+    }
+    
+    /**
+     * Restore cache of Fedex API response.
+     * 
+     * @param \Magento\Fedex\Model\Carrier $subject
+     * @param \Magento\Fedex\Model\Carrier $result
+     * @param \Magento\Quote\Model\Quote\Address\RateRequest $request
+     * @return \Magento\Fedex\Model\Carrier
+     */
+    public function afterSetRequest(\Magento\Fedex\Model\Carrier $subject, $result, $request)
+    {
+        if (!HookHelper::$fromBolt || !$subject->canCollectRates()) {
+            return $result;
+        }
+        $this->restoreFedexResponseFromCacheIfExist($subject, \Magento\Fedex\Model\Carrier::RATE_REQUEST_GENERAL);
+        $this->restoreFedexResponseFromCacheIfExist($subject, \Magento\Fedex\Model\Carrier::RATE_REQUEST_SMARTPOST); 
+        
+        return $result;
+    }
+    
+    /**
+     * Save Fedex API response into cache.
+     * 
+     * @param \Magento\Fedex\Model\Carrier $subject
+     * @param \Magento\Shipping\Model\Rate\Result|bool|null $result
+     * @param \Magento\Quote\Model\Quote\Address\RateRequest $request
+     * @return \Magento\Shipping\Model\Rate\Result|bool|null
+     */
+    public function afterCollectRates(\Magento\Fedex\Model\Carrier $subject, $result, $request)
+    {
+        if (!HookHelper::$fromBolt || !$subject->canCollectRates()) {
+            return $result;
+        }        
+        $this->saveFedexResponseToCacheIfExist($subject, \Magento\Fedex\Model\Carrier::RATE_REQUEST_GENERAL);
+        $this->saveFedexResponseToCacheIfExist($subject, \Magento\Fedex\Model\Carrier::RATE_REQUEST_SMARTPOST);
+        
+        return $result;
+    }
+    
+    /**
+     * Forming request for rate estimation depending to the purpose
+     *
+     * @param \Magento\Fedex\Model\Carrier $subject
+     * @param string $purpose
+     * @return string
+     */
+    private function getFedexRequestString($subject, $purpose)
+    {
+        $ratesRequest = $this->methodCaller->call($subject, '_formRateRequest', $purpose);
+        unset($ratesRequest['RequestedShipment']['ShipTimestamp']);
+        
+        return $this->json->serialize($ratesRequest);
+    }
+    
+    /**
+     * Save Fedex API results into cache.
+     *
+     * @param \Magento\Fedex\Model\Carrier $subject
+     * @param string $purpose
+     */
+    private function saveFedexResponseToCacheIfExist($subject, $purpose)
+    {
+        $requestString = $this->getFedexRequestString($subject, $purpose);
```

Comment:
request string includes both address and cart total, correct?

Reply:
The request string contains address and line item of package, but no cart totals.